### PR TITLE
Make `history --clear` more robust

### PR DESCRIPTION
**Pull request by Hofer-Julian**
- Remove auxiliary files of sqlite
- Give proper error messages if there are problems

Might help with https://github.com/nushell/nushell/issues/7991

### CODE REVIEW

**Comment by Hofer-Julian** [comment]
Ping :)

**Comment by amtoine** [comment]
might supersede https://github.com/nushell/nushell/pull/9575

**Comment by Hofer-Julian** [comment]
@amtoine can this be merged?
This is a clear improvement as far as I can see, and even removes a "TODO" comment.

**Comment by amtoine** [comment]
ooh yeah, i can totally do that :relieved:

File: crates/nu-cli/src/commands/history.rs [hunk 1]
```diff
@@ -64,9 +65,30 @@ impl Command for History {
...
-                Ok(PipelineData::empty())
+                let mut files_to_remove = Vec::from([history_path.clone()]);
+                if let HistoryFileFormat::Sqlite = engine_state.config.history_file_format {
+                    files_to_remove.extend_from_slice(&[
+                        history_dir.join("history.sqlite3-shm"),
+                        history_dir.join("history.sqlite3-wal"),
+                    ])
+                }
+                let (number_errors, message) = files_to_remove
+                    .iter()
```
Comment (amtoine) [inline crates/nu-cli/src/commands/history.rs:76]: do you think there might be more file to remove here?

i mean i kinda like this iterator and the accumulation of the errors to get the nice error i show above :ok_hand: 
but it makes the code a bit more involved at the same time when the iterator only has two elements :yum:

**Review by amtoine — COMMENTED** [review]
that's pretty nice :smirk:
```nushell
> rm ~/.config/nushell/history.sqlite3-*
> history --clear
Error: nu::shell::io_error

  × I/O error
  help: Failed to remove file "/home/amtoine/.config/nushell/history.sqlite3-wal": No such file or directory (os error 2)
        Failed to remove file "/home/amtoine/.config/nushell/history.sqlite3-shm": No such file or directory (os error 2)
```

File: crates/nu-cli/src/commands/history.rs (hunk 1 above)
Comment (Hofer-Julian) [reply crates/nu-cli/src/commands/history.rs:76]: It has either 1 (txt) or 3 (sqlite) files. How would you simplify this? :)

**Comment by amtoine** [comment]
@Hofer-Julian 
have you tried this on Windows?

**Comment by Hofer-Julian** [comment]
> @Hofer-Julian have you tried this on Windows?

I did not.
Is there anything in the code that makes you expect OS specific behaviour?

**Comment by amtoine** [comment]
> I did not. Is there anything in the code that makes you expect OS specific behaviour?
>
yeah... Windows paths :scream: :skull: :laughing:

**Comment by fdncred** [comment]
> Is there anything in the code that makes you expect OS specific behaviour?

Nothing in the code but you have to consider that Windows is different than *nix. Windows does not allow you to mess with things that are in use. You get what are called sharing violations. This is one of the reasons why I originally said that we need to use sql to `delete from history` and `sqlite vacuum` instead of trying to remove the files.
image
and these files still exist
image

**Review by fdncred — COMMENTED** [review]
Doesn't work in Windows so I'm making it a draft.

**Comment by fdncred** [comment]
To add a little more detail, @rgwood tried this before. The PR is buried here some where I'm sure. As I recall, we have to leave 1 record in the database because if you delete everything, nushell tries to get and entry from the db and pukes. So, I'd start with leaving 1 entry. I also think a vacuum is needed. Read more about it here https://www.sqlite.org/lang_vacuum.html.

**Comment by Hofer-Julian** [comment]
> Doesn't work in Windows so I'm making it a draft.

Could you please add the error you are getting. Otherwise this is a bit hard to debug.

**Comment by fdncred** [comment]
> @fdncred I see you started working on this, much appreciated https://github.com/nushell/nushell/pull/11170. Feel free to reach out if you can need help/testing.

#11170 is unrelated to this PR, so I think we still need something to do this.

However, if you want to play around with the `stor` commands, I'd appreciate the feedback and/or changes.

**Comment by AucaCoyan** [comment]
Hello! 
I tried the branch from the repo of Hofer and it worked on my windows machine. It removed all the history but 1 command: `history --clear`.
I cannot reproduce the bug 🤔

**Comment by Hofer-Julian** [comment]
> > @fdncred I see you started working on this, much appreciated #11170. Feel free to reach out if you can need help/testing.
> 
> #11170 is unrelated to this PR, so I think we still need something to do this.
> 
> However, if you want to play around with the `stor` commands, I'd appreciate the feedback and/or changes.

I am confused. "stor reset" sounds exactly like the behavior you are suggesting for "history --clear".

**Comment by Hofer-Julian** [comment]
> Hello! I tried the branch from the repo of Hofer and it worked on my windows machine. It removed all the history but 1 command: `history --clear`. I cannot reproduce the bug 🤔

Did you use the sqlite or file backend?

**Comment by fdncred** [comment]
> I am confused. "stor reset" sounds exactly like the behavior you are suggesting for "history --clear".

The `stor` family of commands are specifically for a sqlite in-memory database that I'm trying to get support for and landed. Those commands do not know anything about history or any sqlite files on the file system.

**Comment by amtoine** [comment]
i'm not sure i understand why this was closed :confused:

**Comment by Hofer-Julian** [comment]
This approach deleted the sqlite files instead of modifying the existing database. Opening a new PR is just as much work as adapting this one. :)

**Comment by amtoine** [comment]
ok, that's fair!

but `history --clear` already removes one file: `history.sqlite3`.
the issue appeared to me to be that `history.sqlite3-shm` and `history.sqlite3-wal` were left over, but you're saying they should all three stay there and be emptied?

there is still the issue that, if you remove the history file yourself, Nushell will crash next time, which imo should still be fixed and was fixed by this very PR :open_mouth: 
```nushell
> rm ~/.config/nushell/history.sqlite3
> nu
Error:   × error within history database: SqliteFailure(Error { code: SystemIoFailure, extended_code: 522 }, Some("disk I/O error"))
```

**Comment by fdncred** [comment]
the shm and wal files are related to allowing sqlite to work concurrently with multiple sessions which is called WAL mode or Write-Ahead-Logging. You can read more about it here https://www.sqlite.org/walformat.html. Once I quit out of using nushell the shm and wal files are flushed into the sqlite history db. It does this on Windows. I haven't checked any other OS. There could be data loss if these files are removed incorrectly.

From sqlite's website: https://www.sqlite.org/tempfiles.html
A write-ahead log or WAL file is used in place of a rollback journal when SQLite is operating in [WAL mode](https://www.sqlite.org/wal.html). As with the rollback journal, the purpose of the WAL file is to implement atomic commit and rollback. The WAL file is always located in the same directory as the database file and has the same name as the database file except with the 4 characters "-wal" appended. The WAL file is created when the first connection to the database is opened and is normally removed when the last connection to the database closes. However, if the last connection does not shutdown cleanly, the WAL file will remain in the filesystem and will be automatically cleaned up the next time the database is opened.


According to Stackoverflow: https://stackoverflow.com/questions/12928294/what-content-do-shm-and-wal-files-contain-after-running-checkpoint
The -shm file does not contain any permanent data.

When the last connection is closed, the database is automatically checkpointed, and the -wal file is then deleted. This implies that after a checkpoint, and if no other connections exist, the -wal file does not contain any important data.

If possible, you should close the connection before taking the backup.

**Pull request closed without merging**